x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageApiKeyConditionalPrivileges.java

```diff
@@ -0,0 +1,224 @@
+        this.realmsPredicate = Automatons.predicate(this.realms);
+        this.users = (users == null) ? Collections.emptySet() : Set.copyOf(users);
+        this.usersPredicate = Automatons.predicate(this.users);
+
+        this.requestPredicate = (request, authentication) -> {
+            if (request instanceof CreateApiKeyRequest && privilege.predicate().test(CREATE_API_KEY_PATTERN)) {
+                return true;
+            } else if (request instanceof GetApiKeyRequest && privilege.predicate().test(GET_API_KEY_PATTERN)) {
+                final GetApiKeyRequest getApiKeyRequest = (GetApiKeyRequest) request;
+                if (this.realms.contains("_self") && this.users.contains("_self")) {
+                    return checkIfUserIsOwnerOfApiKeys(authentication, getApiKeyRequest.getApiKeyId(), getApiKeyRequest.getUserName(),
+                            getApiKeyRequest.getRealmName());
+                } else {
+                    return checkIfAccessAllowed(realms, getApiKeyRequest.getRealmName(), realmsPredicate)
+                            && checkIfAccessAllowed(users, getApiKeyRequest.getUserName(), usersPredicate);
+                }
+            } else if (request instanceof InvalidateApiKeyRequest && privilege.predicate().test(INVALIDATE_API_KEY_PATTERN)) {
+                final InvalidateApiKeyRequest invalidateApiKeyRequest = (InvalidateApiKeyRequest) request;
+                if (this.realms.contains("_self") && this.users.contains("_self")) {
+                    return checkIfUserIsOwnerOfApiKeys(authentication, invalidateApiKeyRequest.getId(),
+                            invalidateApiKeyRequest.getUserName(), invalidateApiKeyRequest.getRealmName());
+                } else {
+                    return checkIfAccessAllowed(realms, invalidateApiKeyRequest.getRealmName(), realmsPredicate)
+                            && checkIfAccessAllowed(users, invalidateApiKeyRequest.getUserName(), usersPredicate);
+                }
+            }
+            return false;
+        };
+    }
+
+    private boolean checkIfUserIsOwnerOfApiKeys(Authentication authentication, String apiKeyId, String username, String realmName) {
+        if (authentication.getAuthenticatedBy().getType().equals("_es_api_key")) {
+            // API key id from authentication must match the id from request
+            String authenticatedApiKeyId = (String) authentication.getMetadata().get("_security_api_key_id");
+            if (Strings.hasText(apiKeyId)) {
+                return apiKeyId.equals(authenticatedApiKeyId);
+            }
+        } else {
+            String authenticatedUserPrincipal = authentication.getUser().principal();
+            String authenticatedUserRealm = authentication.getAuthenticatedBy().getName();
+            if (Strings.hasText(username) && Strings.hasText(realmName)) {
+                return username.equals(authenticatedUserPrincipal) && realmName.equals(authenticatedUserRealm);
+            }
```

Comment:
I don't think this behaviour is what we'd want when `run_as` is used.

Reply:
I don't follow why `run_as` for API keys would come in to play here.

The question is What should the behaviour be if I authenticate as `elastic` and _run-as_ `bizybot` and then try to delete one of `bizybot`'s keys?